### PR TITLE
DFR-3475 update contact details about to submit

### PR DESCRIPTION
**Pull request by al-hmcts**
### Jira link

See [DFR-3475](https://tools.hmcts.net/jira/browse/DFR-3475)

### Change description

Refactor remove-details controller endpoint to be replaced with about to submit handler. 

Note:

Added FinremCaseData overrides in service to allow consented equivalent endpoint to be migrated to FinremCaseData in the future. 

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java [hunk 1]
```diff
@@ -56,6 +59,11 @@ public void persistOrgPolicies(Map<String, Object> caseData, CaseDetails origina
         caseData.put(RESPONDENT_ORGANISATION_POLICY, originalDetails.getData().get(RESPONDENT_ORGANISATION_POLICY));
     }
 
+    public void persistOrgPolicies(FinremCaseData finremCaseData, FinremCaseData originalFinremCaseData) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:62]: Java doc for the public methods?

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java [hunk 2]
```diff
@@ -97,11 +116,38 @@ void removeApplicantSolicitorDetails(Map<String, Object> caseData, String caseTy
         }
     }
 
+    void removeApplicantSolicitorDetails(FinremCaseData caseData, CaseType caseType) {
+        boolean isContested = CaseType.CONTESTED.equals(caseType);
+        ContactDetailsWrapper contactDetailsWrapper = caseData.getContactDetailsWrapper();
+
+        if (contactDetailsWrapper.getApplicantRepresented().isNoOrNull()) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:123]: should check null on `getApplicantRepresented`?

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandler.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        boolean includeRepresentationChange = contactDetailsWrapper.getUpdateIncludesRepresentativeChange().isYes();
+
+        if (includeRepresentationChange) {
+            updateContactDetailsService.handleRepresentationChange(finremCaseData, finremCaseDetails.getCaseType());
+        }
+
+        YesOrNo isRespondentAddressHidden  = contactDetailsWrapper.getRespondentAddressHiddenFromApplicant();
+        YesOrNo isApplicantAddressHidden = contactDetailsWrapper.getApplicantAddressHiddenFromRespondent();
+
+        if ((isRespondentAddressHidden != null && isRespondentAddressHidden.isYes())
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandler.java:70]: `YesOrNo.YES == isRespondentAddressHidden`

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandler.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    @Override
+    public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(FinremCallbackRequest callbackRequest,
+                                                                              String userAuthorisation) {
+        FinremCaseDetails finremCaseDetails = callbackRequest.getCaseDetails();
+        log.info("Invoking contested {} about to submit callback for Case ID: {}",
+            callbackRequest.getEventType(), finremCaseDetails.getId());
+        FinremCaseData finremCaseData = finremCaseDetails.getData();
+
+        ContactDetailsWrapper contactDetailsWrapper = finremCaseData.getContactDetailsWrapper();
+        boolean includeRepresentationChange = contactDetailsWrapper.getUpdateIncludesRepresentativeChange().isYes();
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandler.java:61]: possibly NPE.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java [hunk 5]
```diff
@@ -118,4 +164,29 @@ void removeRespondentDetails(Map<String, Object> caseData, String caseTypeId) {
...
         }
     }
+
+    void removeRespondentDetails(FinremCaseData caseData, CaseType caseType) {
+        boolean isContested = CaseType.CONTESTED.equals(caseType);
+
+        ContactDetailsWrapper contactDetailsWrapper = caseData.getContactDetailsWrapper();
+
+        boolean respondentRepresented = isContested
+            ? contactDetailsWrapper.getContestedRespondentRepresented().isYes()
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:174]: null check

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        request.setCaseDetailsBefore(finremCaseDetailsBefore);
+
+        var response = handler.handle(request, AUTH_TOKEN);
+
+        assertNotNull(response);
+        verify(updateContactDetailsService, times(1)).persistOrgPolicies(finremCaseData, finremCaseDetailsBefore.getData());
+        verify(onlineFormDocumentService, times(1)).generateContestedMiniForm(AUTH_TOKEN, request.getCaseDetails());
+        verify(nocWorkflowService, never()).handleNoticeOfChangeWorkflow(any(), any(), any());
+    }
+
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandlerTest.java:123]: remove an extra blank line.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    void givenNoRepresentationChangeAndRespondentHasHiddenAddresses_handle() {
+
+        FinremCaseData finremCaseData = FinremCaseData.builder().contactDetailsWrapper(
+            ContactDetailsWrapper.builder()
+                .respondentAddressHiddenFromApplicant(YesOrNo.YES)
+                .nocParty(NoticeOfChangeParty.APPLICANT)
+                .updateIncludesRepresentativeChange(YesOrNo.NO)
+                .build()).build();
+
+        FinremCallbackRequest request = FinremCallbackRequestFactory.from(finremCaseData);
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandlerTest.java:90]: pass caseType as well?

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    @Test
+    void givenNoRepresentationChangeAndNoHiddenAddresses_handle() {
+
+        FinremCaseData finremCaseData = FinremCaseData.builder().contactDetailsWrapper(
+            ContactDetailsWrapper.builder()
+                .nocParty(NoticeOfChangeParty.APPLICANT)
+                .updateIncludesRepresentativeChange(YesOrNo.NO)
+                .build()).build();
+
+        FinremCallbackRequest request = FinremCallbackRequestFactory.from(finremCaseData);
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandlerTest.java:69]: pass caseType as well?

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    void givenNoRepresentationChangeAndApplicantHasHiddenAddresses_handle() {
+
+        FinremCaseData finremCaseData = FinremCaseData.builder().contactDetailsWrapper(
+            ContactDetailsWrapper.builder()
+                .applicantAddressHiddenFromRespondent(YesOrNo.YES)
+                .nocParty(NoticeOfChangeParty.APPLICANT)
+                .updateIncludesRepresentativeChange(YesOrNo.NO)
+                .build()).build();
+
+        FinremCallbackRequest request = FinremCallbackRequestFactory.from(finremCaseData);
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandlerTest.java:112]: pass caseType as well?

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandler.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    private final UpdateContactDetailsService updateContactDetailsService;
+    private final OnlineFormDocumentService onlineFormDocumentService;
+    private final UpdateRepresentationWorkflowService nocWorkflowService;
+    private final FinremCaseDetailsMapper detailsMapper;
+
+    public UpdateContactDetailsAboutToSubmitHandler(FinremCaseDetailsMapper finremCaseDetailsMapper,
+                                                    UpdateContactDetailsService updateContactDetailsService,
+                                                    OnlineFormDocumentService onlineFormDocumentService,
+                                                    UpdateRepresentationWorkflowService nocWorkflowService,
+                                                    FinremCaseDetailsMapper detailsMapper) {
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsAboutToSubmitHandler.java:37]: You're injecting this object twice!

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java [hunk 11]
```diff
@@ -97,11 +116,38 @@ void removeApplicantSolicitorDetails(Map<String, Object> caseData, String caseTy
         }
     }
 
+    void removeApplicantSolicitorDetails(FinremCaseData caseData, CaseType caseType) {
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:119]: Can be private

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsServiceTest.java [hunk 12]
```diff
@@ -234,7 +279,241 @@ void shouldHandleRespondentDetails(String caseTypeId, Map<String, Object> caseDa
...
+                .respondentSolicitorFirm("sol firm")
+                .respondentSolicitorAddress(Address.builder().addressLine1("Some Address").build())
+                .respondentSolicitorFirm("0123456789")
+                .respondentSolicitorEmail("some@email.com")
+                .respondentSolicitorDxNumber("DXnumber")
+                .build()
+            ).build();
+    }
+
+    private static FinremCaseData getContestedNotResponentRepresentedFinremCaseData() {
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsServiceTest.java:471]: You can't spell respondent

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsServiceTest.java [hunk 13]
```diff
@@ -234,7 +279,241 @@ void shouldHandleRespondentDetails(String caseTypeId, Map<String, Object> caseDa
...
+
+        service.handleRepresentationChange(finremCaseData, caseType);
+
+        for (Function<FinremCaseData, Object> propertyGetter : propertiesToRemove) {
+            Object value = propertyGetter.apply(finremCaseData);
+            assertNull(value);
+        }
+    }
+
+    public static Stream<Arguments> finremCaseDataParameters() {
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsServiceTest.java:298]: Can be private

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java (hunk 2 above)
Comment (al-hmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:123]: isNoOrNull should handle the null check for us, and prior to this call I have added in and NPE safeguards in the handler for ContactDetailsWrapper.

Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:140]: _isNoOrNull should handle the null check for us_

So if contactDetailsWrapper.getApplicantRepresented == null then what happens when your code is executed?

Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:140]: Those "null" related methods in YesOrNo seem very suspect and misleading.

Comment (al-hmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:140]: I see the error of my ways

Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsService.java:140]: I think you are meant to call them like
`YesOrNo.isNoOrNull(contactDetailsWrapper.getApplicantRepresented())`

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsServiceTest.java (hunk 12 above)
Comment (al-hmcts) [reply src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/UpdateContactDetailsServiceTest.java]: I'm going to down load the OpenDyslexic for intelij now 🤣

**Review by jthmcts — APPROVED** [review]

**Review by so99y — APPROVED** [review]

**Pull request merged**